### ISSUE TITLE
Can't fill container with water of different temperatures

### ISSUE THREAD

**Issue by PiscotOficial** [CleverRaven/Cataclysm-DDA#45471]
Game version:
Build #11168

Operating system:
Windows 10 64-bit

Tiles or curses:
Tiles

Mods active:
Disable NPC Needs
No Fungal Monsters
SpeedyDex
Stats Through Skills

### Expected behavior
Container fills with two types of water and perhaps averages the temperatures.

### Actual behavior
Option to unload water into container does not show up if container has different temperature water than what you're trying to unload.

### Steps to reproduce the behavior
Fill any portable container partially with water (cold).
Attempting to further fill it with anything but water (cold) is impossible.

**Comment by Hirmuolio** [comment]
At least with the (U) Unload menu the option is there.

Pouring water (cold) into bottle with normal water: 
![unload](https://user-images.githubusercontent.com/22011552/99429036-1e547880-2910-11eb-8758-57ee540cdaec.png)

**Comment by wapcaplet** [comment]
Confirmed in 0.E-7587-g8d87f156b9 - however, I am not convinced the temperature has anything to do with it.

In my inventory I have some `clean water (cold)` in a `gallon jug`, and some `clean water` (not cold) in a plastic bottle:

![image](https://user-images.githubusercontent.com/238395/99460498-2fd85900-28ed-11eb-8a11-95f4de84373d.png)

The gallon jug has only one 0.25L serving of `clean water (cold)`, plenty of room for more:

![image](https://user-images.githubusercontent.com/238395/99460680-85ad0100-28ed-11eb-9898-a8dfd439c0c3.png)

When I select the `clean water (plastic bottle)` and `U`nload, choosing to "Pour into a container":

![image](https://user-images.githubusercontent.com/238395/99460592-5bf3da00-28ed-11eb-8e4b-a1b7d3a2d689.png)

The only container I can put it in is my empty plastic bottle. The gallon jug of clean water is not available, despite having enough remaining volume:

![image](https://user-images.githubusercontent.com/238395/99460622-6910c900-28ed-11eb-9d43-66fbfb3fb64d.png)

However, if I wait an hour until all the water is the same temperature, I still cannot `U`nload the plastic bottle's water into the gallon jug. And if I put them both in a refrigerator until they are both `(cold)`, even then I cannot combine them into the gallon jug.

**Comment by Hirmuolio** [comment]
If you enable debug mode and look at the water are there any extra tags, variables or other differences on them?

**Comment by wapcaplet** [comment]
The attached savegame has a reproduction case. Player has 1 sealed plastic bottle of clean water, and 2 partially-full gallon jugs of clean water:

![image](https://user-images.githubusercontent.com/238395/99465283-45eb1700-28f7-11eb-85f0-2caad6b2226b.png)

If you attempt to `U`nload the sealed plastic bottle, neither of the gallon jugs is eligible for filling, though they have room.

Using debug mode, I can confirm that the water in the gallon jugs is `EATEN_COLD` and `NO_REPAIR`, while the water in the plastic bottle has neither of those flags.

Savegame .zip: [water-unload.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/5556908/water-unload.zip)

**Comment by sacrificerXY** [comment]
Possibly related #45191

**Comment by PiscotOficial** [comment]
So the problem is that naturally generated liquids are different from crafted ones?

**Comment by sacrificerXY** [comment]
Yeah. The game asks if the 2 liquids can be stacked together. One condition it checks are the flags of the 2 items:
https://github.com/CleverRaven/Cataclysm-DDA/blob/c764d46aac54e9bf2a7a7d52ab244373e6cdd876/src/item.cpp#L972-L978
Here you can see it compares all the flags except the flags for hot, cold, or frozen. Since the debug water has extra flags, this returns false.

So to fix this, either modify this code. Or change the way debug spawning works.

**Comment by wapcaplet** [comment]
After some investigation, I see that the debug-menu spawned item instances have all the item type flags explicitly added to them since #44675. The instance flags can be removed by pressing `F` before spawning the item, and deleting the flags:

![image](https://user-images.githubusercontent.com/238395/99863595-21dc3f80-2b5c-11eb-8e4f-4df73bc4ce30.png)

When testing with liquids spawned in this way, they work as intended - hot and cold water can be mixed in the same container, since they do not have any conflicting instance flags.

Liquids spawned in the game world - water from a puddle, clean water from a water heater tank, mead or moonshine from a barrel - all correctly have no instance "tags" shown (since the flags are inherited from the parent item type), and in all of my testing so far, can be mixed in the same container regardless of temperature.

In other words, I believe now this is working as intended. Would like to know if anyone can reproduce this issue using only in-game spawned liquids, or with cleared instance flags using `F` from the debug item spawning screen.

**Comment by jbytheway** [comment]
It's arguably a bug that it's not clear to the player why these liquids cannot be mixed.

**Comment by wapcaplet** [comment]
I have reproduced this in 0.E-8509-g35c0af53e0 using this attached saved game: [CleanWaterNotMixing.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/5757793/CleanWaterNotMixing.zip)

![image](https://user-images.githubusercontent.com/238395/103418825-98059600-4b4d-11eb-82fb-f8fbc5933984.png)

This character has 2 gallon jugs in the inventory - one of them contains 10 units of room-temperature clean water (crafted a couple days before), and the other contains 3 units of hot clean water (just crafted). Neither jug can be poured into the other using `i`nsert or `U`nload. If you `r`eload either of the jugs, the message log says "You refill the gallon jug > clean water", but no liquid is transferred.

With debugging view turned on, I can see that the clean water in both these jugs has no additional tags except `HOT` for the hot water, so it should be allowed to combine these two jugs of clean water.

**Comment by PaulBlay** [comment]
As far as I know, the latest information is that the temperature is irrelevant, the flags are irrelevant but clean water produced by heating it has 'component' information and will not mix with water that does not have 'component' information. The component in question is that clean water is made from COOKED water.

However, under some unknown circumstances, the component information may be lost. Sometimes. If that happens can combine them.

**Comment by BrettDong** [comment]
> I have reproduced this in 0.E-8509-g35c0af53e0 using this attached saved game: CleanWaterNotMixing.zip

Room temperature clean water has no `components` info while hot clean water has `components` info and that's causing they refuse to stack in item.cpp line 1056 in `item::stacks_with()`:
```c++
    if( check_components || is_comestible() || is_craft() ) {
        //Only check if at least one item isn't using the default recipe or is comestible
        if( !components.empty() || !rhs.components.empty() ) {
            if( get_uncraft_components() != rhs.get_uncraft_components() ) {
                return false;
```

```
(lldb) p this->get_uncraft_components()
(std::vector<item_comp, std::allocator<item_comp> >) $11 = size=0 {}
(lldb) p rhs.get_uncraft_components()
(std::vector<item_comp, std::allocator<item_comp> >) $12 = size=1 {
  [0] = {
    component = {
      type = {
        _version = -1
        _cid = -1
        _id = (_id = 178, _string_id = "water")
      }
      count = 1
      available = a_false
      recoverable = true
      requirement = false
    }
  }
}
```

**Comment by BrettDong** [comment]
This is the same issue as in #44791.
